### PR TITLE
Laboratory work #3, Anastasiia Seliverstova - 23FPL1

### PR DESCRIPTION
**Pull request by seliverstovaa**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -183,6 +247,18 @@ def vector2tokens(self, vector: Vector) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if len(vector) != len(self._vocabulary):
+            return None
+        self.build()
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:252]: it isn't necessary

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -220,6 +296,28 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if (not isinstance(document, list) or len(document) == 0
+                or not all(isinstance(word, str) for word in document)):
+            return None
+        self.build()
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:302]: .

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Good work! I'm waiting for your implementation on 8

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -49,6 +51,16 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if (query_vector is None or document_vector is None
+            or not all(isinstance(elem, float) for elem in query_vector)
+            or not all(isinstance(elem, float) for elem in document_vector)):
+        return None
+    if not query_vector or not document_vector:
+        return 0.0
+    distance = m.sqrt(sum((query_vector[index] - document_vector[index])**2
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:60]: fixed

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -157,6 +199,23 @@ def build(self) -> bool:
         Returns:
             bool: True if built successfully, False in other case
         """
+        if (not isinstance(self._corpus, list) or not self._corpus
+                or not all(isinstance(tokens, list) for tokens in self._corpus)
+                or not all(isinstance(word, str) for tokens in self._corpus for word in tokens)):
+            return False
+        self._vocabulary = []
+        self._idf_values = {}
+        self._token2ind = {}
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:208]: done

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:252]: done

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:302]: done

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -170,6 +229,11 @@ def vectorize(self, tokenized_document: list[str]) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(tokenized_document, list) or len(tokenized_document) == 0\
+                or not all(isinstance(word, str) for word in tokenized_document):
+            return None
+        vector = self._calculate_tf_idf(tokenized_document)
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:235]: without it some tests are failed

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:302]: Some tests don't work without it

File: lab_3_ann_retriever/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:235]: done

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -104,6 +117,19 @@ def tokenize(self, text: str) -> list[str] | None:
...
+        if not isinstance(text, str):
+            return None
+        tokens = []
+        text = text.lower()
+        for elem in text:
+            if not elem.isalpha():
+                elem = elem.replace(elem, ' ')
+            tokens.append(elem)
+        tokens = ''.join(tokens).split(' ')
+        while '' in tokens:
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:128]: what exactly?

**Review by Vasilisa-Blyudova — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -149,6 +185,11 @@ def __init__(self, corpus: list[list[str]]) -> None:
         Args:
             corpus (list[list[str]]): Tokenized documents to vectorize
         """
+        self._corpus = corpus
+        self._vocabulary = []
+        self._idf_values = {}
+        self._token2ind = {}
+        self.build()
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:192]: do yoy really build here?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Very good work! Go to 10!

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:192]: without it there is an error in test_vectorizer_return_value

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -49,6 +53,15 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if (query_vector is None or document_vector is None
+            or not all(isinstance(elem, float) for elem in query_vector)
+            or not all(isinstance(elem, float) for elem in document_vector)):
+        return None
+    if not query_vector or not document_vector:
+        return 0.0
+    distance = math.sqrt(sum((query_vector[index] - document_vector[index])**2
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:62]: just return, without the variable

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -61,6 +74,14 @@ def save_vector(vector: Vector) -> dict:
     Returns:
         dict: A state of the vector to save
     """
+    index_value = {}
+    for index, value in enumerate(vector):
+        if value == 0.0:
+            continue
+        index_value[index] = value
+    result = {"len": len(vector),
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:82]: the same

File: lab_3_ann_retriever/main.py [hunk 10]
```diff
@@ -104,6 +136,14 @@ def tokenize(self, text: str) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+        text = text.lower()
+        for elem in text:
+            if not elem.isalpha() and elem != ' ':
+                text = text.replace(elem, ' ')
+        clear_tokens = self._remove_stop_words(text.lower().split())
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:145]: again

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (artyomtugaryov) [reply lab_3_ann_retriever/main.py:211]: strange. You must not run build in the init

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py (hunk 8 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:62]: fixed

File: lab_3_ann_retriever/main.py (hunk 9 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:82]: done

File: lab_3_ann_retriever/main.py (hunk 10 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:145]: done

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (seliverstovaa) [reply lab_3_ann_retriever/main.py:211]: agree

**Pull request closed without merging**